site/app/libraries/database/DatabaseQueries.php

```diff
@@ -722,6 +722,76 @@ public function getCategoriesIdForThread($thread_id) {
         return $categories_list;
     }
 
+    public function toggleLikes(int $post_id, int $thread_id, string $current_user): string {
+        try {
+            $this->course_db->query("SELECT * FROM forum_upducks WHERE post_id = ? AND user_id = ? AND thread_id = ?", [$post_id, $current_user, $thread_id]);
+
+            if (isset($this->course_db->rows()[0])) {
+                // If a row with the same values exists, delete it
+                $this->course_db->query("DELETE FROM forum_upducks WHERE post_id = ? AND user_id = ? AND thread_id = ?", [$post_id, $current_user, $thread_id]);
+                $result = 'unlike';
+            }
+            else {
+                // If no row with the same values exists, insert the new row
+                $this->course_db->query("INSERT INTO forum_upducks (post_id, user_id, thread_id) VALUES (?, ?, ?)", [$post_id, $current_user, $thread_id]);
+                $result = 'like';
+            }
+            return $result;
+        }
+        catch (DatabaseException $dbException) {
+            if ($this->course_db->inTransaction()) {
+                $this->course_db->rollback();
+            }
+            return "false";
+        }
+    }
+
+    /**
+     * @param int[] $post_ids
+     * @return int[]
+     */
+    public function getUpduckInfoForPosts(array $post_ids): array {
+        $placeholders = $this->createParameterList(count($post_ids));
+        $sql = "SELECT post_id, COUNT(*) AS cnt FROM forum_upducks WHERE post_id IN {$placeholders} GROUP BY post_id";
```

Comment:
Queries look good and not iterated.

One thing I recently had to do with my queries is add a check for an empty post_ids array, and not executing the query if there are no post_ids, as trying to query an empty list will cause a server error (see #10055). Can you add similar checks here and to getUserLikesForPosts?